internal/keystore/vault/client.go

```diff
@@ -155,55 +159,66 @@ func (c *client) RenewToken(ctx context.Context, authenticate authFunc, secret *
 		// another go routine to unblock this for-loop
 		// once vault becomes unsealed again.
 		if c.Sealed() {
-			timer := time.NewTimer(1 * time.Second)
 			select {
 			case <-ctx.Done():
-				if !timer.Stop() {
-					<-timer.C
-				}
 				return
-			case <-timer.C:
+			case <-time.After(time.Second):
 			}
 			continue
 		}
 
-		// We renew the token after 80% of its TTL has passed.
-		renewIn := 80 * (ttl / 100)
-		if renewIn < time.Second {
-			renewIn = time.Second
+		// If the token is about to expire, we re-auth immediately. We also don't wait
+		// for Vault not to sync up because the existing token will become invalid.
+		if ttl < Delay {
+			if s, _ = authenticate(ctx); s != nil {
+				ttl, _ = s.TokenTTL()
+				token, _ := s.TokenID()
+				c.SetToken(token) // SetToken is safe to call from different go routines
+			} else {
+				select {
+				case <-ctx.Done():
+					return
+				case <-time.After(3 * time.Second): // In case of an auth failure, wait 3s and retry
+				}
```

Comment:
I think let's sleep in all cases and not only when there is an error. If someone configures a low TTL < Delay, then it can enter a loop of authentication without sleep (I would say extreme use case, but well..)

Reply:
Yeah, right - sleeping in all cases now